### PR TITLE
front: stdcm: linked train search improvements

### LINKED ISSUE #9545: stdcm: refacto and improvements for the linked path
### Description and goal

We merge last week this PR to add the linked paths. However, we can improve a bit the code:
- [ ] deactivate the path linked cards when needed (if the simulation is running or if the user has selected a simulation)
- [ ] app crashes when I click on the datePicker to open the calendar
- [ ] if I'm on mode debug and I try to fetch a train from a custom scenario, it does not appear. This is because my trains do not have a schedule on their last step. Actually, we should not get the final arrival time of the train from its schedule, but we should get it using the summary endpoint
- [ ] in the simulation report sheet, should we still display the placeholders for the posterior and anterior trains ?
- [ ] we use everywhere linked path, but I think that linked train should be more appropriate ?*

Refacto ideas
- [ ] the component `StdcmLinkedPathSearch` should only take a type of linked train: ANTERIOR or POSTERIOR (we can name it LinkedTrainType). From this information, we can deduce all the other props
- [ ] `StdcmLinkedPathResults` does not need the linked Op if it has the LinkedTrainType. It can simplify a lot the code

### Acceptance criteria

The code related to the Linked Paths is cleaner

### LINKED ISSUE #9618: lmr: deleting a linked path and re-opening a new one keeps track of the previous input
### What happened?

I'm looking for a linked path, then deleting it, and when looking for another linked path, the previous input is still here:
Image

Image

Image

### What did you expect to happen?

Deleting the linked path should clear this component completely. When re-opening it it should be like this:
Image

If a linked path was already selected and the `Origine` fileds were pre-filled, they should be cleared too when the linked path is deleted.

### How can we reproduce it (as minimally and precisely as possible)?

1. Look for a linked path
2. Delete it
3. Look for another one

### On which environments the bug occurs?

Dev (SNCF)

### On which browser the bug occurs?

Google Chrome

### OSRD version (top right corner `Account` button > `Informations`)

pr 9602 (merge of fam/fix-user-identity@1e971f6 into dev@bfa5205)

### LINKED ISSUE #9619: lmr: dates aren't consistent between `Origine` and `Sillon antérieur` and create some weird behaviors
### What happened?

The available dates in the `Origine` datepicker are Sept 26th and 27th
Image

But in the `Sillon antérieur` datepicker i can select Spetember 25th
Image

And if i do so, the search is acting weird
Image
Image
Image

The loader appears for less than a second then I get no error message.

### What did you expect to happen?

I think matching the dates and not having Sept 25th here would fix the bug.

### How can we reproduce it (as minimally and precisely as possible)?

1. Go to lmr on dev environment

### On which environments the bug occurs?

Dev (SNCF)

### On which browser the bug occurs?

Google Chrome

### OSRD version (top right corner `Account` button > `Informations`)

pr 9602 (merge of fam/fix-user-identity@1e971f6 into dev@bfa5205)

### PR DESCRIPTION
**Pull request by SarahBellaha**
Closes #9618 
Closes #9619 
Partially closes #9545

- App no longer crashes when using the datePicker on debug mode.
- We now get the final arrival time using the summary endpoint.
- The op search payload has been adapted so that it can be searched by `obj_id` or by `uic` code and `ch` code.

### CODE REVIEW

File: front/src/applications/stdcm/utils/computeOpSchedules.ts [hunk 1]
```diff
@@ -19,12 +19,12 @@ const computeOpSchedules = (startTime: string, secondsFromStartTime: string) =>
     origin: {
       date: originDate,
       time: originTime,
-      isoArrivalTime: substractDurationToIsoDate(startTime, 'PT1800S'),
+      isoArrivalTime: substractDurationToIsoDate(startTime, 1800),
```
Comment (Uriel-Sautron) [inline front/src/applications/stdcm/utils/computeOpSchedules.ts:36]: Maybe a comment to explain the 1800s

File: front/src/applications/stdcm/hooks/useLinkedPathSearch.ts [hunk 2]
```diff
@@ -32,28 +39,28 @@ const useLinkedPathSearch = () => {
...
-        const groupedResults = groupBy(resultsOP, 'obj_id');
-        return {
-          origin: groupedResults[originId][0],
-          destination: groupedResults[destinationId][0],
+          query: pathItemQuery,
         };
+        const opDetails = await postSearch({ searchPayload: payloadOP, pageSize: 25 }).unwrap();
+        return opDetails[0];
       } catch (error) {
         console.error('Failed to fetch operational points:', error);
```
Comment (Uriel-Sautron) [inline front/src/applications/stdcm/hooks/useLinkedPathSearch.ts]: Typo: operational point

**Review by Uriel-Sautron — APPROVED** [review]
LGTM, tested

File: front/src/applications/stdcm/hooks/useLinkedPathSearch.ts (hunk 2 above)
Comment (SarahBellaha) [reply front/src/applications/stdcm/hooks/useLinkedPathSearch.ts]: Done, thx :)

File: front/src/applications/stdcm/utils/computeOpSchedules.ts (hunk 1 above)
Comment (SarahBellaha) [reply front/src/applications/stdcm/utils/computeOpSchedules.ts:36]: Done, i commented the whole function :)

File: front/src/applications/stdcm/hooks/useLinkedPathSearch.ts [hunk 3]
```diff
@@ -135,6 +135,13 @@ const useLinkedPathSearch = () => {
     }
   }, [postSearch, trainNameInput, timetableId, linkedPathDate, getExtremitiesDetails]);
 
+  useEffect(() => {
+    if (!isArrivalDateInSearchTimeWindow(linkedTrainDate, searchDatetimeWindow)) {
+      setLinkedTrainDate(selectableSlot.start);
+      launchTrainScheduleSearch();
```
Comment (clarani) [inline front/src/applications/stdcm/hooks/useLinkedPathSearch.ts]: Why do you launch the search here ? I think it would be better to reset the bool hasSearchBeenLaunched and the results, what do you think ?

File: front/src/utils/date.ts [hunk 4]
```diff
@@ -218,18 +218,18 @@ export function extractDateAndTimefromISO(arrivalTime: string, dateFormat: strin
...
  * Checks if the given arrival date falls within the specified search time window.
  *
- * @param {string} arrivalTime - The arrival time as a string, which will be parsed into a Date object.
+ * @param {string | Date} arrivalTime - The arrival time, which can be a Date object or an ISO string that will be parsed into a Date object.
  * @param {{ begin: Date; end: Date } | undefined} searchDatetimeWindow - An object containing the start and end dates of the search window. If undefined, the function will return true.
  * @returns {boolean} - Returns true if the arrival date is within the search time window, or if the search time window is undefined. Returns false otherwise.
  */
 export function isArrivalDateInSearchTimeWindow(
-  arrivalTime: string,
+  arrivalTime: string | Date,
```
Comment (clarani) [inline front/src/utils/date.ts]: I think it would be better to transform this function so it can only takes a date, can you add a commit to do this ?

File: front/src/applications/stdcm/components/StdcmForm/StdcmOperationalPoint.tsx [hunk 5]
```diff
@@ -82,6 +82,7 @@ const StdcmOperationalPoint = ({ point, opPointId, disabled }: StdcmOperationalP
           ch: p.ch,
           uic: p.uic,
           coordinates: p.geographic.coordinates,
+          trigram: p.trigram,
```
Comment (clarani) [inline front/src/applications/stdcm/components/StdcmForm/StdcmOperationalPoint.tsx]: Why did you add the trigram here ? I think it's not in the type :/ (here we use the ItemLocationVariant with the uic)

File: front/src/applications/stdcm/hooks/useLinkedPathSearch.ts [hunk 6]
```diff
@@ -35,28 +39,28 @@ const useLinkedPathSearch = () => {
...
-          query: ['or', ['=', ['obj_id'], originId], ['=', ['obj_id'], destinationId]],
-        };
-        const resultsOP = await postSearch({ searchPayload: payloadOP, pageSize: 25 }).unwrap();
-        const groupedResults = groupBy(resultsOP, 'obj_id');
-        return {
-          origin: groupedResults[originId][0],
-          destination: groupedResults[destinationId][0],
+          query: pathItemQuery,
         };
+        const opDetails = await postSearch({ searchPayload: payloadOP, pageSize: 25 }).unwrap();
```
Comment (clarani) [inline front/src/applications/stdcm/hooks/useLinkedPathSearch.ts]: You can add a cast here 
```suggestion
        const opDetails = (await postSearch({
          searchPayload: payloadOP,
          pageSize: 25,
        }).unwrap()) as SearchResultItemOperationalPoint[];
```

File: front/src/applications/stdcm/hooks/useLinkedPathSearch.ts [hunk 7]
```diff
@@ -108,20 +114,19 @@ const useLinkedPathSearch = () => {
...
+          const originDetails = await getExtremityDetails(result.path.at(0)!);
+          const destinationDetails = await getExtremityDetails(result.path.at(-1)!);
           const computedOpSchedules = computeOpSchedules(result.start_time, msFromStartTime);
 
-          if (opDetails === undefined) return undefined;
+          if (!originDetails || !destinationDetails) return undefined;
           return {
             trainName: result.train_name,
-            origin: { ...opDetails.origin, ...computedOpSchedules.origin } as StdcmLinkedPathStep,
+            origin: { ...originDetails, ...computedOpSchedules.origin } as StdcmLinkedPathStep,
```
Comment (clarani) [inline front/src/applications/stdcm/hooks/useLinkedPathSearch.ts]: If you have added the cast in getExtremityDetails, you can remove this one
```suggestion
            origin: { ...originDetails, ...computedOpSchedules.origin },
```

File: front/src/applications/stdcm/hooks/useLinkedPathSearch.ts [hunk 8]
```diff
@@ -133,11 +138,11 @@ const useLinkedPathSearch = () => {
...
       setDisplaySearchButton(true);
     }
-  }, [postSearch, trainNameInput, timetableId, linkedPathDate, getExtremitiesDetails]);
+  }, [postSearch, trainNameInput, timetableId, linkedPathDate, getExtremityDetails]);
 
   useEffect(() => {
-    if (!isArrivalDateInSearchTimeWindow(linkedTrainDate, searchDatetimeWindow)) {
-      setLinkedTrainDate(selectableSlot.start);
+    if (!isArrivalDateInSearchTimeWindow(linkedPathDate, searchDatetimeWindow)) {
+      setLinkedPathDate(selectableSlot.start);
```
Comment (clarani) [inline front/src/applications/stdcm/hooks/useLinkedPathSearch.ts:153]: These changes should be in your 2nd commit I think (fyi the 2nd commit does not build :/)

File: front/src/applications/stdcm/components/StdcmForm/StdcmLinkedPathSearch.tsx [hunk 9]
```diff
@@ -99,12 +99,12 @@ const StdcmLinkedPathSearch = ({
...
             </button>
           )}
-          {!displaySearchButton && !linkedPathResults.length && (
+          {!displaySearchButton && !linkedPathResults && (
             <div className="stdcm-linked-path-button white">
               <Gear size="lg" className="stdcm-linked-path-loading" />
             </div>
           )}
-          {linkedPathResults.length > 0 ? (
+          {linkedPathResults && linkedPathResults.length > 0 ? (
```
Comment (clarani) [inline front/src/applications/stdcm/components/StdcmForm/StdcmLinkedPathSearch.tsx]: I'm not sure but I think you can remove the boolean hasSearchBeenLaunched ?

File: front/src/applications/stdcm/hooks/useLinkedPathSearch.ts [hunk 10]
```diff
@@ -140,6 +140,13 @@ const useLinkedPathSearch = () => {
     }
   }, [postSearch, trainNameInput, timetableId, linkedPathDate, getExtremityDetails]);
 
+  const resetLinkedPathSearch = () => {
```
Comment (clarani) [inline front/src/applications/stdcm/hooks/useLinkedPathSearch.ts:145]: Nice !

File: front/src/applications/stdcm/hooks/useStdcmEnv.tsx [hunk 11]
```diff
@@ -23,14 +24,16 @@ export default function useStdcmEnvironment() {
       setLoading(true);
       const { data } = await getStdcmSearchEnvironment();
       if (!data) throw new Error(NO_CONFIG_FOUND_MSG);
+      // We want to extract the date from the ISO string to avoid timezone issues
+      const { arrivalDate } = extractDateAndTimefromISO(data.search_window_begin);
```
Comment (clarani) [inline front/src/applications/stdcm/hooks/useStdcmEnv.tsx]: Why do you do this only for the begin and not also for the end ?

**Review by Uriel-Sautron — APPROVED** [review]
Tested, good work.

**Review by clarani — COMMENTED** [review]
I left some of my previous comments open :)

File: front/src/applications/stdcm/hooks/useStdcmEnv.tsx (hunk 11 above)
Comment (clarani) [reply front/src/applications/stdcm/hooks/useStdcmEnv.tsx]: I think there is a bug in editoast if the date is not iso. We should not convert the date here, I think you can remove these lines.
If there is a bug in editoast, we should fix it there :+1:

**Review by clarani — APPROVED** [review]
LGTM :white_check_mark: 

Can you squash all your commits ?

**Review by Maymanaf — APPROVED** [review]
LGTM for e2e tests

**Review by clarani — APPROVED** [review]
LGTM :white_check_mark:

**Review by clarani — APPROVED** [review]
LGTM :white_check_mark:
